File: app/src/androidTest/java/com/star_zero/example/androidmvvm/matcher/InputTextLayoutMatcher.java
package com.star_zero.example.androidmvvm.matcher;

import android.content.res.Resources;
import android.support.design.widget.TextInputLayout;
import android.support.test.espresso.matcher.BoundedMatcher;
import android.view.View;

import org.hamcrest.Description;
import org.hamcrest.Matcher;

public class InputTextLayoutMatcher {

    public static Matcher<View> isErrorEnabled(final boolean enabled) {
        return new BoundedMatcher<View, TextInputLayout>(TextInputLayout.class) {
            @Override
            public void describeTo(Description description) {
                description.appendText("isErrorEnabled: " + enabled);
            }

            @Override
            public boolean matchesSafely(TextInputLayout view) {
                return enabled == view.isErrorEnabled();
            }
        };
    }

    public static Matcher<View> withErrorText(final int resId) {
        return new BoundedMatcher<View, TextInputLayout>(TextInputLayout.class) {
            @Override
            public void describeTo(Description description) {
                description.appendText("error resource id: " + resId);
            }

            @Override
            public boolean matchesSafely(TextInputLayout view) {
                String expect;
                try {
                    expect = view.getResources().getString(resId);
                } catch (Resources.NotFoundException e) {
                    return false;
                }

                String actual = view.getError().toString();

                return expect.equals(actual);
            }
        };
    }
}

File: app/src/androidTest/kotlin/com/star_zero/example/androidmvvm/matcher/InputTextLayoutMatcher.kt
package com.star_zero.example.androidmvvm.matcher

import android.content.res.Resources
import android.support.design.widget.TextInputLayout
import android.support.test.espresso.matcher.BoundedMatcher
import android.view.View

import org.hamcrest.Description
import org.hamcrest.Matcher

object InputTextLayoutMatcher {

    fun isErrorEnabled(enabled: Boolean): Matcher<View> {
        return object : BoundedMatcher<View, TextInputLayout>(TextInputLayout::class.java) {
            override fun describeTo(description: Description) {
                description.appendText("isErrorEnabled: " + enabled)
            }

            public override fun matchesSafely(view: TextInputLayout): Boolean {
                return enabled == view.isErrorEnabled
            }
        }
    }

    fun withErrorText(resId: Int): Matcher<View> {
        return object : BoundedMatcher<View, TextInputLayout>(TextInputLayout::class.java) {
            override fun describeTo(description: Description) {
                description.appendText("error resource id: " + resId)
            }

            public override fun matchesSafely(view: TextInputLayout): Boolean {
                val expect: String
                try {
                    expect = view.resources.getString(resId)
                } catch (e: Resources.NotFoundException) {
                    return false
                }

                val actual = view.error!!.toString()

                return expect == actual
            }
        }
    }
}
